File: 2021/day2/scala/solution.scala
import scala.io.Source
enum Direction:
  case forward, up, down

case class Move(direction: Direction, units: Int)

case class Position(horizontal : Int, depth: Int, aim: Int = 0)

@main
def main() = {
  println("Part 1: ")
  print("Example: ")
  part1("../example.txt")
  print("Input: ")
  part1("../input.txt")
  println()

  println("Part 2: ")
  print("Example: ")
  part2("../example.txt")
  print("Input: ")
  part2("../input.txt")
  println()
}

def parseInput(path: String) = {
  Source.fromFile(path)
  .getLines
  .map((line: String) => {
    val Array(direction, units) = line.split(" ")
    Move(Direction.valueOf(direction), units.toInt)
  })
  .toList
}

def part1(path: String) = {
  val start = Position(0, 0)

  val stop = parseInput(path)
  .foldLeft(start)((position, nextMove) => {
    nextMove match {
      case Move(Direction.forward, n) => position.copy(position.horizontal + n)
      case Move(Direction.down, n) => position.copy(depth = position.depth + n)
      case Move(Direction.up, n) => position.copy(depth = position.depth - n)
    }
  })
  print(s"$stop = ")
  println(stop.depth * stop.horizontal)
}

def part2(path: String) = {
  val start = Position(0, 0, 0)

  val stop = parseInput(path)
  .foldLeft(start)((position, nextMove) => {
    nextMove match {
      case Move(Direction.forward, n) => position.copy(
        horizontal = position.horizontal + n,
        depth = position.depth + position.aim * n
      )
      case Move(Direction.down, n) => position.copy(aim = position.aim + n)
      case Move(Direction.up, n) => position.copy(aim = position.aim - n)
    }
  })
  print(s"$stop = ")
  println(stop.depth * stop.horizontal)
}

File: 2021/day2/ruby/solution.rb
input = File.readlines(__dir__ + '/../input.txt')
example = File.readlines(__dir__ + '/../example.txt')

def parse_command(line)
  command, value = line.split(' ')
  [command, value.to_i]
end

def part1(input)
  resulting_position = input
    .map(&method(:parse_command))
    .each_with_object({ depth: 0, horizontal: 0 }) do |cmd, position|
    direction, value = cmd
    case direction
    in "forward" then position[:horizontal] += value
    in "down" then position[:depth] += value
    in "up" then position[:depth] -= value
    end
  end

  resulting_position.values.reduce(&:*)
end

def part2(input)
  resulting_position = input
     .map(&method(:parse_command))
     .each_with_object({ depth: 0, horizontal: 0, aim: 0 }) do |cmd, position|
    direction, value = cmd
    case direction
    in "forward" then
      position[:horizontal] += value
      position[:depth] += position[:aim] * value
    in "down" then position[:aim] += value
    in "up" then position[:aim] -= value
    end
  end

  resulting_position.values_at(:depth, :horizontal).reduce(&:*)
end

puts "Example: #{part1(example)}"
puts "Part 1: #{part1(input)}"
puts "Part 2: #{part2(input)}"
